### PR TITLE
Fix IDGenerator may generate zero TraceId / SpanId 

### LINKED ISSUE #5462: IDGenerator may generate zero TraceId / SpanId
### Description

The spec (https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md) mentions that TraceId and SpanId must have at least one non-zero byte. However, the default id generator doesn't ensure that (https://github.com/open-telemetry/opentelemetry-go/blob/main/sdk/trace/id_generator.go#L44). The generator simply uses rand.Read() which does not guarantee non-zero output.

### Environment

All environments

### Steps To Reproduce

Difficult to give exact steps to reproduce, as random number generation is involved.

### Expected behavior

Generated TraceId and SpanId must always be non-zero as per spec.

### PR DESCRIPTION
**Pull request by Charlie-lizhihan**
# Description
Fix #5462 

## Type of change

add a loop to generate the spanID and traceID. 
the loop will not stop until it generate a valid ID

- [x] Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -30,6 +30,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 
 ### Fixed
 
+- Fix IDGenerator may generate zero TraceId / SpanId. (#5514)
```
Comment (dmathieu) [inline CHANGELOG.md:33]:
```suggestion
- Retry trace and span ID generation if it generated an invalid one in `go.opentelemetry.io/otel/sdk/trace`. (#5514)
```

**Review by dmathieu — COMMENTED** [review]
Could you run the benchmarks for the trace SDK, and provide their results before and after this change? (using [benchstat](https://pkg.go.dev/golang.org/x/perf/cmd/benchstat) to provide them in a comparable format).

File: CHANGELOG.md (hunk 1 above)
Comment (Charlie-lizhihan) [reply CHANGELOG.md]: Thanks for this!

**Comment by Charlie-lizhihan** [comment]
> Could you run the benchmarks for the trace SDK, and provide their results before and after this change? (using [benchstat](https://pkg.go.dev/golang.org/x/perf/cmd/benchstat) to provide them in a comparable format).

@dmathieu Hi Damien, I run the banchstat and got result like this:

```
% benchstat old_bench.txt new_bench.txt

goos: darwin
goarch: arm64
pkg: go.opentelemetry.io/otel/sdk/trace
                                             │ old_bench.txt │            new_bench.txt             │
                                             │    sec/op     │    sec/op     vs base                │
RecordingSpanSetAttributes/WithLimit/false-8    3.663µ ± ∞ ¹   3.267µ ± ∞ ¹       ~ (p=1.000 n=1) ²
RecordingSpanSetAttributes/WithLimit/true-8    10.180µ ± ∞ ¹   9.377µ ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanProcessor-8                                 12.24µ ± ∞ ¹   12.02µ ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanProcessorVerboseLogging-8                   13.98µ ± ∞ ¹   11.98µ ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanLimits/AttributeValueLengthLimit-8          5.744µ ± ∞ ¹   5.640µ ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanLimits/AttributeCountLimit-8                5.797µ ± ∞ ¹   5.129µ ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanLimits/EventCountLimit-8                    4.946µ ± ∞ ¹   4.978µ ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanLimits/LinkCountLimit-8                     4.913µ ± ∞ ¹   4.925µ ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanLimits/AttributePerEventCountLimit-8        5.280µ ± ∞ ¹   5.323µ ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanLimits/AttributePerLinkCountLimit-8         5.240µ ± ∞ ¹   5.270µ ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanSetAttributesOverCapacity-8                 1.937µ ± ∞ ¹   1.947µ ± ∞ ¹       ~ (p=1.000 n=1) ²
StartEndSpan/AlwaysSample-8                     906.4n ± ∞ ¹   906.5n ± ∞ ¹       ~ (p=1.000 n=1) ²
StartEndSpan/NeverSample-8                      179.6n ± ∞ ¹   184.6n ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithAttributes_4/AlwaysSample-8             1.162µ ± ∞ ¹   1.168µ ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithAttributes_4/NeverSample-8              283.1n ± ∞ ¹   287.0n ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithAttributes_8/AlwaysSample-8             1.360µ ± ∞ ¹   1.357µ ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithAttributes_8/NeverSample-8              359.2n ± ∞ ¹   363.2n ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithAttributes_all/AlwaysSample-8           1.228µ ± ∞ ¹   1.231µ ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithAttributes_all/NeverSample-8            307.8n ± ∞ ¹   310.6n ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithAttributes_all_2x/AlwaysSample-8        1.484µ ± ∞ ¹   1.480µ ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithAttributes_all_2x/NeverSample-8         404.6n ± ∞ ¹   409.6n ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithEvents_4/AlwaysSample-8                 1.410µ ± ∞ ¹   1.417µ ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithEvents_4/NeverSample-8                  187.5n ± ∞ ¹   191.4n ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithEvents_8/AlwaysSample-8                 1.894µ ± ∞ ¹   1.901µ ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithEvents_8/NeverSample-8                  192.1n ± ∞ ¹   197.8n ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithEvents_WithStackTrace/AlwaysSample-8    1.048µ ± ∞ ¹   1.067µ ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithEvents_WithStackTrace/NeverSample-8     205.0n ± ∞ ¹   211.7n ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithEvents_WithTimestamp/AlwaysSample-8     1.040µ ± ∞ ¹   1.046µ ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithEvents_WithTimestamp/NeverSample-8      237.8n ± ∞ ¹   242.7n ± ∞ ¹       ~ (p=1.000 n=1) ²
TraceID_DotString-8                             46.27n ± ∞ ¹   46.08n ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanID_DotString-8                              35.68n ± ∞ ¹   41.35n ± ∞ ¹       ~ (p=1.000 n=1) ²
geomean                                         1.088µ         1.084µ        -0.41%
¹ need >= 6 samples for confidence interval at level 0.95
² need >= 4 samples to detect a difference at alpha level 0.05

                                             │ old_bench.txt │             new_bench.txt             │
                                             │     B/op      │     B/op       vs base                │
RecordingSpanSetAttributes/WithLimit/false-8   7.379Ki ± ∞ ¹   7.379Ki ± ∞ ¹       ~ (p=1.000 n=1) ²
RecordingSpanSetAttributes/WithLimit/true-8    11.54Ki ± ∞ ¹   11.54Ki ± ∞ ¹       ~ (p=1.000 n=1) ³
SpanProcessor-8                                14.06Ki ± ∞ ¹   14.06Ki ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanProcessorVerboseLogging-8                  14.05Ki ± ∞ ¹   14.05Ki ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanLimits/AttributeValueLengthLimit-8         10.29Ki ± ∞ ¹   10.29Ki ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanLimits/AttributeCountLimit-8               9.584Ki ± ∞ ¹   9.584Ki ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanLimits/EventCountLimit-8                   9.161Ki ± ∞ ¹   9.161Ki ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanLimits/LinkCountLimit-8                    8.833Ki ± ∞ ¹   8.833Ki ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanLimits/AttributePerEventCountLimit-8       10.15Ki ± ∞ ¹   10.15Ki ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanLimits/AttributePerLinkCountLimit-8        10.15Ki ± ∞ ¹   10.15Ki ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanSetAttributesOverCapacity-8                1.062Ki ± ∞ ¹   1.062Ki ± ∞ ¹       ~ (p=1.000 n=1) ²
StartEndSpan/AlwaysSample-8                    1.000Ki ± ∞ ¹   1.000Ki ± ∞ ¹       ~ (p=1.000 n=1) ²
StartEndSpan/NeverSample-8                       144.0 ± ∞ ¹     144.0 ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithAttributes_4/AlwaysSample-8            1.500Ki ± ∞ ¹   1.500Ki ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithAttributes_4/NeverSample-8               400.0 ± ∞ ¹     400.0 ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithAttributes_8/AlwaysSample-8            2.001Ki ± ∞ ¹   2.001Ki ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithAttributes_8/NeverSample-8               656.0 ± ∞ ¹     656.0 ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithAttributes_all/AlwaysSample-8          1.626Ki ± ∞ ¹   1.626Ki ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithAttributes_all/NeverSample-8             464.0 ± ∞ ¹     464.0 ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithAttributes_all_2x/AlwaysSample-8       2.251Ki ± ∞ ¹   2.251Ki ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithAttributes_all_2x/NeverSample-8          784.0 ± ∞ ¹     784.0 ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithEvents_4/AlwaysSample-8                1.500Ki ± ∞ ¹   1.500Ki ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithEvents_4/NeverSample-8                   144.0 ± ∞ ¹     144.0 ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithEvents_8/AlwaysSample-8                2.063Ki ± ∞ ¹   2.063Ki ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithEvents_8/NeverSample-8                   144.0 ± ∞ ¹     144.0 ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithEvents_WithStackTrace/AlwaysSample-8   1.094Ki ± ∞ ¹   1.094Ki ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithEvents_WithStackTrace/NeverSample-8      160.0 ± ∞ ¹     160.0 ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithEvents_WithTimestamp/AlwaysSample-8    1.117Ki ± ∞ ¹   1.117Ki ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithEvents_WithTimestamp/NeverSample-8       184.0 ± ∞ ¹     184.0 ± ∞ ¹       ~ (p=1.000 n=1) ²
TraceID_DotString-8                              64.00 ± ∞ ¹     64.00 ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanID_DotString-8                               32.00 ± ∞ ¹     32.00 ± ∞ ¹       ~ (p=1.000 n=1) ²
geomean                                        1.337Ki         1.337Ki        +0.00%
¹ need >= 6 samples for confidence interval at level 0.95
² all samples are equal
³ need >= 4 samples to detect a difference at alpha level 0.05

                                             │ old_bench.txt │            new_bench.txt            │
                                             │   allocs/op   │  allocs/op   vs base                │
RecordingSpanSetAttributes/WithLimit/false-8     19.00 ± ∞ ¹   19.00 ± ∞ ¹       ~ (p=1.000 n=1) ²
RecordingSpanSetAttributes/WithLimit/true-8      25.00 ± ∞ ¹   25.00 ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanProcessor-8                                  195.0 ± ∞ ¹   195.0 ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanProcessorVerboseLogging-8                    195.0 ± ∞ ¹   195.0 ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanLimits/AttributeValueLengthLimit-8           59.00 ± ∞ ¹   59.00 ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanLimits/AttributeCountLimit-8                 54.00 ± ∞ ¹   54.00 ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanLimits/EventCountLimit-8                     51.00 ± ∞ ¹   51.00 ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanLimits/LinkCountLimit-8                      51.00 ± ∞ ¹   51.00 ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanLimits/AttributePerEventCountLimit-8         54.00 ± ∞ ¹   54.00 ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanLimits/AttributePerLinkCountLimit-8          54.00 ± ∞ ¹   54.00 ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanSetAttributesOverCapacity-8                  19.00 ± ∞ ¹   19.00 ± ∞ ¹       ~ (p=1.000 n=1) ²
StartEndSpan/AlwaysSample-8                      18.00 ± ∞ ¹   18.00 ± ∞ ¹       ~ (p=1.000 n=1) ²
StartEndSpan/NeverSample-8                       2.000 ± ∞ ¹   2.000 ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithAttributes_4/AlwaysSample-8              20.00 ± ∞ ¹   20.00 ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithAttributes_4/NeverSample-8               3.000 ± ∞ ¹   3.000 ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithAttributes_8/AlwaysSample-8              20.00 ± ∞ ¹   20.00 ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithAttributes_8/NeverSample-8               3.000 ± ∞ ¹   3.000 ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithAttributes_all/AlwaysSample-8            20.00 ± ∞ ¹   20.00 ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithAttributes_all/NeverSample-8             3.000 ± ∞ ¹   3.000 ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithAttributes_all_2x/AlwaysSample-8         20.00 ± ∞ ¹   20.00 ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithAttributes_all_2x/NeverSample-8          3.000 ± ∞ ¹   3.000 ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithEvents_4/AlwaysSample-8                  21.00 ± ∞ ¹   21.00 ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithEvents_4/NeverSample-8                   2.000 ± ∞ ¹   2.000 ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithEvents_8/AlwaysSample-8                  22.00 ± ∞ ¹   22.00 ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithEvents_8/NeverSample-8                   2.000 ± ∞ ¹   2.000 ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithEvents_WithStackTrace/AlwaysSample-8     20.00 ± ∞ ¹   20.00 ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithEvents_WithStackTrace/NeverSample-8      3.000 ± ∞ ¹   3.000 ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithEvents_WithTimestamp/AlwaysSample-8      21.00 ± ∞ ¹   21.00 ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanWithEvents_WithTimestamp/NeverSample-8       4.000 ± ∞ ¹   4.000 ± ∞ ¹       ~ (p=1.000 n=1) ²
TraceID_DotString-8                              2.000 ± ∞ ¹   2.000 ± ∞ ¹       ~ (p=1.000 n=1) ²
SpanID_DotString-8                               2.000 ± ∞ ¹   2.000 ± ∞ ¹       ~ (p=1.000 n=1) ²
geomean                                          13.62         13.62        +0.00%
¹ need >= 6 samples for confidence interval at level 0.95
² all samples are equal
```

**Comment by Charlie-lizhihan** [comment]
> Could you run the benchmarks for the trace SDK, and provide their results before and after this change? (using [benchstat](https://pkg.go.dev/golang.org/x/perf/cmd/benchstat) to provide them in a comparable format).

Hello! This is my first time submitting a PR, so I have a couple of questions:

After finishing my code, I understand I should run make test and make precommit, followed by make codespell. Are there any additional checks I need to perform before submitting my PR?

Regarding the benchmark request, should I include benchstat results in every pull request, or is it specific to this one?

Thank you for your guidance!

**Comment by dmathieu** [comment]
> After finishing my code, I understand I should run make test and make precommit, followed by make codespell. Are there any additional checks I need to perform before submitting my PR?

CI runs them as well (and they passed). So you're good here.

> Regarding the benchmark request, should I include benchstat results in every pull request, or is it specific to this one?

For things that may impact performance only.

You will find more information about contributing here: https://github.com/open-telemetry/opentelemetry-go/blob/main/CONTRIBUTING.md

**Review by dmathieu — APPROVED** [review]

File: sdk/trace/id_generator.go [hunk 2]
```diff
@@ -51,9 +55,19 @@ func (gen *randomIDGenerator) NewIDs(ctx context.Context) (trace.TraceID, trace.
 	gen.Lock()
 	defer gen.Unlock()
 	tid := trace.TraceID{}
-	_, _ = gen.randSource.Read(tid[:])
 	sid := trace.SpanID{}
-	_, _ = gen.randSource.Read(sid[:])
+	for {
+		_, _ = gen.randSource.Read(tid[:])
+		if tid.IsValid() {
+			break
```
Comment (dmathieu) [inline sdk/trace/id_generator.go:62]: Same pattern here and below?

Comment (Charlie-lizhihan) [reply sdk/trace/id_generator.go:62]: should be good now, thanks!

**Review by dmathieu — APPROVED** [review]

File: sdk/trace/id_generator.go [hunk 3]
```diff
@@ -51,10 +55,19 @@ func (gen *randomIDGenerator) NewIDs(ctx context.Context) (trace.TraceID, trace.
...
+		}
+	}
+	for {
+		_, _ = gen.randSource.Read(sid[:])
+		if sid.IsValid() {
+			return tid, sid
+		}
+	}
```
Comment (pellared) [inline sdk/trace/id_generator.go:70]:
```suggestion
		if sid.IsValid() {
			break
		}
	}
	return tid, sid
```

File: sdk/trace/id_generator.go [hunk 4]
```diff
@@ -41,8 +41,12 @@ func (gen *randomIDGenerator) NewSpanID(ctx context.Context, traceID trace.Trace
...
 	defer gen.Unlock()
 	sid := trace.SpanID{}
-	_, _ = gen.randSource.Read(sid[:])
-	return sid
+	for {
+		_, _ = gen.randSource.Read(sid[:])
+		if sid.IsValid() {
+			return sid
+		}
+	}
```
[the remark this answers is not in the document]
Comment (pellared) [reply sdk/trace/id_generator.go:49]: IMO having `return` not indented for "regular" cases is more idiomatic. Based on https://go.dev/wiki/CodeReviewComments#indent-error-flow

**Review by pellared — APPROVED** [review]

**Review by MrAlias — APPROVED** [review]

**Pull request merged**